### ISSUE TITLE
segfault on mailmime_write_* on specific mailheaders

### ISSUE THREAD

**Issue by loomy** [dinhviethoa/libetpan#205]
Hi,

I encountered a segfault if your trying to write out a mail using the mailmime_write functions if the mail includes an empty Header Value. 
in my example

    Return-Path: <>

parsing this mailheader with

    mailimf_message_parse

then attaching the mailimf fields to the first mime

    mailmime_set_imf_fields(mailcontainer, mailimf_msg->msg_fields)

everything is fine until you try to write out the mail. valgrindoutput:

    ==13577== Invalid read of size 1
    ==13577==    at 0x4029C38: strlen (mc_replace_strmem.c:399) 
    ==13577==    by 0x41E9190: mailimf_field_write_driver (mailimf_write_generic.c:1120)
    ==13577==    by 0x41E9518: mailimf_fields_write_driver (mailimf_write_generic.c:768)
    ==13577==    by 0x41F71B2: mailmime_part_write_driver (mailmime_write_generic.c:939)
    ==13577==    by 0x41F77F4: mailmime_write_mem (mailmime_write_mem.c:92)
    ==13577==    by 0x804A1E1: bosxml_2_mimemail (boslta-restore.c:311)
    ==13577==    by 0x804976E: main (boslta-restore.c:593)
    ==13577==  Address 0x0 is not stack'd, malloc'd or (recently) free'd

tested against the newest git sources.
in a quick fix I added a check in the specific write_driver

this works for me and seems logical. maybe it would be a good idea to add this kind of check in some other write functions too?

    

<pre>
diff --git a/src/low-level/imf/mailimf_write_generic.c b/src/low-level/imf/mailimf_write_generic.c
index 23af6ce..17a101f 100644
--- a/src/low-level/imf/mailimf_write_generic.c
+++ b/src/low-level/imf/mailimf_write_generic.c
@@ -1773,10 +1773,12 @@ static int mailimf_path_write_driver(int (* do_write)(void *, const char *, size
   if (r != MAILIMF_NO_ERROR)
     return r;
 
-  r = mailimf_string_write_driver(do_write, data, col, path->pt_addr_spec,
-      strlen(path->pt_addr_spec));
-  if (r != MAILIMF_NO_ERROR)
-    return r;
+  if (path->pt_addr_spec && strlen(path->pt_addr_spec)){
+      r = mailimf_string_write_driver(do_write, data, col, path->pt_addr_spec,
+          strlen(path->pt_addr_spec));
+      if (r != MAILIMF_NO_ERROR)
+        return r;
+  }
 
   r = mailimf_string_write_driver(do_write, data, col, ">", 1);
   if (r != MAILIMF_NO_ERROR)
</pre>

**Comment by dinhviethoa** [comment]
Would this work?
```
 if (path->pt_addr_spec != NULL) {
```

instead of:
```
if (path->pt_addr_spec && strlen(path->pt_addr_spec)) {
```

**Comment by loomy** [comment]
hi,

yepp. works the same way. 

which is strange to me. I checked the pointer on earlier tests and it was just an empty char * as it should be.
thats why I added the strlen check. doesnt make much sense to write out an empty string anyway.

**Comment by dinhviethoa** [comment]
Fixed with 2bbd57231f3f349a9c22b7f3eac3cf9d31476744.
